src/index.ts

```diff
@@ -23,108 +39,203 @@ export {
-			stream.on('end', next);
 		},
 	);
 
-	extract.on('finish', async () => {
-		// Ensure that at least one resolver is satisfied, otherwise emit an error
-
-		// Firstly, check that if the user specified a dockerfile, and that dockerfile has
-		// been found and processed
-		let resolver: Resolver;
-
-		if (dockerfile != null) {
-			if (specifiedFileResolver == null) {
-				pack.emit(
-					'error',
-					new Error(
-						`Specified dockerfile could not be resolved: ${dockerfile}`,
-					),
-				);
-				return;
-			}
-			resolver = specifiedFileResolver;
-		} else {
-			// Detect if any of the resolvers have been satisfied
-			const satisfied = _(resolvers)
-				.filter(r => r.isSatisfied(bundle))
-				.orderBy('priority', 'desc')
-				.value();
-
-			if (satisfied.length === 0) {
-				pack.emit('error', new Error('Resolution could not be performed'));
-				return;
-			}
+	extract.once('finish', async () => {
+		try {
+			console.log('resin-bundle-resolve extract on finish');
 
-			resolver = satisfied[0];
-			try {
-				await addResolverOutput(bundle, resolver, pack);
-			} catch (e) {
-				pack.emit('error', e);
-				return;
+			await resolveTarStreamOnFinish(
+				bundle,
+				resolvers,
+				pack,
+				specifiedFileResolver,
+				dockerfile,
+			);
+		} catch (error) {
+			console.log('resin-bundle-resolve extract finish -> error', error);
+			pack.emit(error);
+		} finally {
+			console.log(
+				'resin-bundle-resolve extract on finish calling pack.finalize()',
+			);
+			pack.finalize();
+		}
+	});
+
+	bundle.tarStream.pipe(extract);
+	return pack;
+}
+
+async function resolveTarStreamOnEntry(
+	entry: { header: tar.Headers; stream: Readable; next: () => void },
+	bundle: Bundle,
+	resolvers: Resolver[],
+	pack: tar.Pack,
+	specifiedFileResolver: null | Resolver = null,
+	dockerfile?: string,
+): Promise<Resolver | undefined> {
+	const name = Utils.normalizeTarEntry(entry.header.name);
+
+	// Handle the case where the dockerfile is specified
+	if (dockerfile != null) {
+		if (specifiedFileResolver == null && name === dockerfile) {
+			specifiedFileResolver = await resolveSpecifiedFile(
+				resolvers,
+				bundle,
+				name,
+				entry.header,
+				entry.stream,
+				pack,
+			);
+			entry.next();
+			return specifiedFileResolver;
+		}
+	} else {
+		const potentials = resolvers.filter(r => r.needsEntry(name));
+		if (potentials.length > 0) {
+			const fileInfo = await streamToFileInfo(entry.stream, entry.header);
+			for (const resolver of potentials) {
+				resolver.entry(fileInfo);
 			}
+			// Also add it to the stream
+			pack.entry(entry.header, fileInfo.contents);
+			entry.next();
+			return;
 		}
+	}
+	const newEntryStream = pack.entry(entry.header);
+	const newEntryStreamPromise = new Promise((resolve, reject) => {
+		newEntryStream.on('error', reject);
+		newEntryStream.on('finish', () => {
+			entry.next();
+			resolve();
+		});
+	});
+	entry.stream.pipe(newEntryStream);
+	await newEntryStreamPromise;
```

Comment:
The ~10 lines above (`newEntryStreamPromise`) were necessary to avoid a `tar-stream` limitation by which it is only possible to pipe to one pack.entry at a time -- so it's necessary to wait for the 'finish' event on one entry before moving on to the next entry. More details here: https://github.com/mafintosh/tar-stream/issues/24#issuecomment-54120650